### ISSUE TITLE
Additional package publishing support to nimble

### ISSUE THREAD

**Issue by samdmarshall** [nim-lang/nimble#622]
**Scenario:**
Due to the ease-of-use of nimble and nim (overall, as a language), I've ended up with a bunch of small packages/programs that i wrote to solve trivial problems in my development environment and shell config (checking email, calendar, contacts, etc). As a result I have a bunch of repos of nimble-packages that i use daily. Originally I kept making package formulae to install them via homebrew/linuxbrew (my package manager of choice) but this has grown really tedious and only requires a lot more boiler-plate be maintained for the sake of just myself. So, instead of going through the cycle of "making local changes -> push to GitHub -> reinstall package from HEAD via brew", i've switch to creating a local package list for nimble instead.

**Friction:**
Overall this has been pretty good, it required I make some changes (for the better) to some of the nimble packages I have to handle hybrid packages the correct way. The major source of friction here is maintaining my local custom packages list. This requires manually writing in new json for each package I want to have access to and sync across workstations with my dot-files so I can ensure that my working environment is always the same. I really don't like writing json.

**Proposal:**
Adding support to `nimble` to:

1. publish a package to any package list known to `nimble` (assuming it isn't read-only/published via a platform with no interface to make changes)
2. add optional meta-data to the `.nimble` file that specifies the package-list to publish too
3. handle publishing a package to a package-list that doesn't require forking to make a contribution to (eg: a package-list repo on the same account as the access token is for (don't try to fork your own repo))
4. better define the steps of publishing a package (due to cases outside of the official package list) and the requirements/expectations a user should have when performing the `publish` command.

---

I realize that this covers a lot of stuff and there is probably a lot I haven't considered about this yet I think the major/top-level points are covered here and hashing out the details of each scenario/use-case can be done as this work gets implemented. I wanted to see if my proposal here aligns with your road-map and how best to approach implementing these features.

**Comment by Araq** [comment]
You could publish this is a small article, it's a nice read and we are crying for more publicity. :-)

**Comment by samdmarshall** [comment]
@Araq, should i take that as an affirmative that the proposed feature of nimble publishing to other package lists (besides the official channel) as on the road-map and a contribution i could attempt here? 🙂🤔

**Comment by Araq** [comment]
Yes from me but @dom96 is in charge.

**Comment by dom96** [comment]
> publish a package to any package list known to nimble (assuming it isn't read-only/published via a platform with no interface to make changes)

I would be happy with ``nimble publish <git_repo_with_package_json_here>`` and/or ``nimble publish <packages_name_from_nimbles_config_file>``

> add optional meta-data to the .nimble file that specifies the package-list to publish too

Don't think this is warranted.

> handle publishing a package to a package-list that doesn't require forking to make a contribution to (eg: a package-list repo on the same account as the access token is for (don't try to fork your own repo))

AFAIK this is actually a bug that needs to be fixed.

> better define the steps of publishing a package (due to cases outside of the official package list) and the requirements/expectations a user should have when performing the publish command.

👍

**Comment by samdmarshall** [comment]
Cool! then i'll get started on this tomorrow!

**Comment by samdmarshall** [comment]
Oops, my mistake, pressing the wrong buttons. 

@dom96 i'm looking over the code now and i'm wondering how much the current style of the command parsing is desired. The impression i'm getting is that this started small and features got added and code grew as needed -- rather than starting out with a strong foundation of parsing commands? From an architecture stand-point it seems like there are a couple of basic flags that would alter global values (eg: `--nimbleDir:`, `--noColor`, etc.) and then commands (`build`, `install`, `publish`, etc.) then there are flags on those commands (eg: `list --installed`, `search --ver`, `install --depsOnly`, etc.). However the whole parsing mechanism feels like it is all-over-the-place instead of resolved into individual units (like, parse up until the first command, then the main OptParser stops, and that specific command's OptParser takes the remaining unparsed args via `.cmdLineRest()`/`.remainingArgs()` and parses that to resolve the commands.

I ask this because the core of this issue is not just the superficial "make `nimble publish …` go other places", but rather allow nimble to interact with sources (package lists) in a more user-friendly way -- and that requires changing a lot more code than just `publish.nim`, as it would have to establish naming convention and mechanism to handle interacting with individual package lists throughout the whole program.

**Comment by dom96** [comment]
Huh? All flags are parsed into a data structure called `Options` (https://github.com/nim-lang/nimble/blob/master/src/nimblepkg/options.nim#L10). I don't see how the parsing is "all over the place".

To answer your question: I'd say the current style is desired, but maybe I'm missing something.

**Comment by samdmarshall** [comment]
Okay, I think i came off as overly-hostile in that previous comment and that wasn't my intention, so: I apologize for not being more considerate in my word-choice.

I think much of my confusion arose from the ambiguity of the various flags and parameters that are accepted. When it comes to code for mapping inputs to intended actions, my preference is towards defining all inputs of a thing, for example:
```
type Options = object
  case cmd: SubCommand
  of scInstall:
    installPkgs: seq[string]
    onlyDependencies: bool      # representation of `nimble install [-d|--depsOnly] ... `
  …
  of scSearch:
    searchPattern: string
    searchShowVersions: bool    # representation of `nimble search --ver … `
  of scList:
    listShowVersions: bool      # representation of `nimble list --ver … `
  …
```
instead of having it summarized as a top-level flag; which, to me, seems to indicate that `nimble --ver …` is a valid input, but actually isn't. So, my previous comment did jump-the-gun a bit in terms of how the inputs get represented on the command line. however, i still feel that the concluding statement, "this is more than just parsing names of package lists", is true. This seems to have the potential of much greater complexity than it seems at face value to me, but maybe i'm misunderstanding this -- as other package managers I've worked on have had to deal with a much larger scope than the current nim-lang community -- so I might be ballooning the potential complexity of this in my own mind more-so than it will in reality. Either way, I feel a bit unsure about just implementing this with such little discussion on it that has potential to have a major impact in usage (perhaps i'm just overly cautious).

**Comment by dom96** [comment]
I see what you mean and yes, I agree, it would be clearer to have the flags inside the ``Action`` structure. You can refactor this but for the sake of resolving this issue I don't think it will be a problem.

**Comment by samdmarshall** [comment]
yeah, I don't want to get hung up on refactoring that when it isn't the main goal of this issue; I just wanted to make sure what I was leaving behind would be something that accommodates other goals as well.
